### ISSUE TITLE
Consider command-line option for GHC backend to not generate unreachable catch-alls

### ISSUE THREAD

**Issue by WolframKahl** [agda/agda#1902]
I just did an experiment, and removed the `_ -> MAlonzo.RTE.mazUnreachableError` catch-alls: The resulting executable allocates exactly the same number of bytes, but runs 5% to 7% faster.

```
--- src/full/Agda/Compiler/MAlonzo/Compiler.hs~ 2016-03-03 09:51:10.000000000 -0500
+++ src/full/Agda/Compiler/MAlonzo/Compiler.hs  2016-03-11 21:20:05.008592575 -0500
@@ -424,9 +424,12 @@
     sc' <- term (T.TVar sc)
     alts' <- traverse (alt sc) alts
     def' <- term def
-    let defAlt = HS.Alt dummy HS.PWildCard (HS.UnGuardedRhs def') emptyBinds
+    let defAlt = case def of
+          -- 2016-03-11 WK: Uncommenting the following line gives 5% speed-up:
+          -- T.TError T.TUnreachable -> []  -- This needs  -fno-warn-incomplete-patterns !
+          _ -> (: []) $ HS.Alt dummy HS.PWildCard (HS.UnGuardedRhs def') emptyBinds

-    return $ HS.Case (hsCast sc') (alts' ++ [defAlt])
+    return $ HS.Case (hsCast sc') (alts' ++ defAlt)

   T.TLit l -> lift $ literal l
   T.TDef q -> do
```

**Comment by phile314** [comment]
The semantics of `TError TUnreachable` is undefined behaviour(!), so applying your patch as-is sounds good to me. We don't need a command line option for that in my opinion.

The default case could only be helpful if we want to provide better error messages at some point like <pre>Error: Got unexpected constructor ... in case expression in file ...</pre> But we are not doing that right now anyway, and this would be better handled by generating a different catchall along the lines of `TError TImpossibleCase` or so.

**Comment by phile314** [comment]
<pre>
data TError
  = TUnreachable
  -- ^ Code which is unreachable. E.g. absurd branches or missing case defaults.
  -- Runtime behaviour of unreachable code is undefined, but preferably
  -- the program will exit with an error message. The compiler is free
  -- to assume that this code is unreachable and to remove it.
  deriving (Typeable, Show, Eq, Ord)
</pre>

**Comment by WolframKahl** [comment]
Yes. But at least for the GHC backend, applying this patch makes a change in the call to GHC necessary, as I documented in the patch:

```
-- This needs  -fno-warn-incomplete-patterns !
```

(It may also change the experience of debugging Agda compiler errors, so perhaps you may want to put the current default behaviour into a command-line option `--ghc-verbose-unreachable-branches`.)

**Comment by UlfNorell** [comment]
The problem is that we do use the `-fwarn-incomplete-patterns` to ensure that `COMPILED_DATA` types have the right constructors. For instance, if you try something like

``` agda
data Nil (A : Set) : Set where
  [] : Nil A

{-# COMPILED_DATA Nil [] #-}
```

you'll get an error (from ghc) complaining that you're missing a cons case. I suppose it might be possible to put these checks in separate modules...

**Comment by WolframKahl** [comment]
The easy way would be to document that the new option should only be used after compilation without it passes GHC. And to also document that failures are only expected from `COMPILED_DATA` and from Agda compiler errors.

**Comment by jespercockx** [comment]
Is there perhaps a way to turn `-fwarn-incomplete-patterns` on/off locally in the generated haskell code?

**Comment by UlfNorell** [comment]
Not that I'm aware of.
